ntcore/src/main/native/cpp/LocalStorage.h

```diff
@@ -106,117 +126,183 @@ class LocalStorage final : public net::ILocalStorage {
+    if (auto topic = m_impl.GetTopicByHandle(topicHandle)) {
+      return m_impl.SetProperties(topic, update, true);
+    } else {
+      return {};
+    }
+  }
 
   TopicInfo GetTopicInfo(NT_Topic topicHandle) {
     std::scoped_lock lock{m_mutex};
-    if (auto topic = m_impl.m_topics.Get(topicHandle)) {
+    if (auto topic = m_impl.GetTopicByHandle(topicHandle)) {
       return topic->GetTopicInfo();
     } else {
       return {};
     }
   }
 
-  NT_Subscriber Subscribe(NT_Topic topic, NT_Type type,
+  NT_Subscriber Subscribe(NT_Topic topicHandle, NT_Type type,
                           std::string_view typeStr,
-                          const PubSubOptions& options);
+                          const PubSubOptions& options) {
+    std::scoped_lock lock{m_mutex};
+    if (auto topic = m_impl.GetTopicByHandle(topicHandle)) {
+      if (auto subscriber = m_impl.Subscribe(topic, type, typeStr, options)) {
+        return subscriber->handle;
+      }
+    }
+    return {};
+  }
 
   void Unsubscribe(NT_Subscriber subHandle) {
     std::scoped_lock lock{m_mutex};
     m_impl.RemoveSubEntry(subHandle);
   }
 
   NT_MultiSubscriber SubscribeMultiple(
-      std::span<const std::string_view> prefixes, const PubSubOptions& options);
+      std::span<const std::string_view> prefixes,
+      const PubSubOptions& options) {
+    std::scoped_lock lock{m_mutex};
+    if (auto sub = m_impl.AddMultiSubscriber(prefixes, options)) {
+      return sub->handle;
+    } else {
+      return {};
+    }
+  }
 
   void UnsubscribeMultiple(NT_MultiSubscriber subHandle) {
     std::scoped_lock lock{m_mutex};
     m_impl.RemoveMultiSubscriber(subHandle);
   }
 
-  NT_Publisher Publish(NT_Topic topic, NT_Type type, std::string_view typeStr,
-                       const wpi::json& properties,
-                       const PubSubOptions& options);
+  NT_Publisher Publish(NT_Topic topicHandle, NT_Type type,
+                       std::string_view typeStr, const wpi::json& properties,
+                       const PubSubOptions& options) {
+    std::scoped_lock lock{m_mutex};
+    if (auto topic = m_impl.GetTopicByHandle(topicHandle)) {
+      if (auto publisher =
+              m_impl.Publish(topic, type, typeStr, properties, options)) {
+        return publisher->handle;
+      }
+    } else {
+      WPI_ERROR(m_impl.GetLogger(),
+                "trying to publish invalid topic handle ({})", topicHandle);
+    }
+    return {};
+  }
 
-  void Unpublish(NT_Handle pubentry);
+  void Unpublish(NT_Handle pubentryHandle) {
+    std::scoped_lock lock{m_mutex};
+    if (Handle{pubentryHandle}.IsType(Handle::kPublisher)) {
+      m_impl.RemoveLocalPublisher(pubentryHandle);
+    } else if (auto entry = m_impl.GetEntryByHandle(pubentryHandle)) {
+      if (entry->publisher) {
+        m_impl.RemoveLocalPublisher(entry->publisher->handle);
+        entry->publisher = nullptr;
+      }
+    } else {
+      // TODO: report warning
+      return;
+    }
+  }
```

Comment:
Would it be worth it to move the impl into `LocalStorageImpl` so that `LocalStorageImpl::RemoveLocalPublisher` could be made private like `AddLocalPublisher`, `RemoveLocalSubscriber`, and `AddLocalSubscriber`?